FirebaseAppDistribution/Sources/FIRAppDistribution.m

```diff
@@ -0,0 +1,270 @@
+        completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+          if (error) {
+            // TODO: Reformat error into error code
+            completion(nil, error);
+            return;
+          }
+
+          NSHTTPURLResponse *HTTPResponse = (NSHTTPURLResponse *)response;
+
+          if (HTTPResponse.statusCode == 200) {
+            [self handleReleasesAPIResponseWithData:data completion:completion];
+          } else {
+            // TODO: Handle non-200 http response
+            @throw([NSException exceptionWithName:@"NotImplementedException"
+                                           reason:@"This code path is not implemented yet"
+                                         userInfo:nil]);
+          }
+        }];
+
+  [listReleasesDataTask resume];
+}
+
+- (void)handleOauthDiscoveryCompletion:(OIDServiceConfiguration *_Nullable)configuration
+                                 error:(NSError *_Nullable)error
+       appDistributionSignInCompletion:(void (^)(NSError *_Nullable error))completion {
+  if (!configuration) {
+    // TODO: Handle when we cannot get configuration
+    @throw([NSException exceptionWithName:@"NotImplementedException"
+                                   reason:@"This code path is not implemented yet"
+                                 userInfo:nil]);
+    return;
+  }
+
+  NSString *redirectURL = [@"dev.firebase.appdistribution."
+      stringByAppendingString:[[[NSBundle mainBundle] bundleIdentifier]
+                                  stringByAppendingString:@":/launch"]];
+
+  OIDAuthorizationRequest *request = [[OIDAuthorizationRequest alloc]
+      initWithConfiguration:configuration
+                   clientId:kTesterAPIClientID
+                     scopes:@[ OIDScopeOpenID, OIDScopeProfile, kOIDScopeTesterAPI ]
+                redirectURL:[NSURL URLWithString:redirectURL]
+               responseType:OIDResponseTypeCode
+       additionalParameters:nil];
+
+  [self createUIWindowForLogin];
+  // performs authentication request
+  [FIRAppDistributionAppDelegatorInterceptor sharedInstance].currentAuthorizationFlow =
+      [OIDAuthState authStateByPresentingAuthorizationRequest:request
+                                     presentingViewController:self.safariHostingViewController
+                                                     callback:^(OIDAuthState *_Nullable authState,
+                                                                NSError *_Nullable error) {
+                                                       self.authState = authState;
+                                                       self.isTesterSignedIn =
+                                                           self.authState ? YES : NO;
+                                                       completion(error);
+                                                     }];
+}
+
+- (UIWindow *)createUIWindowForLogin {
+  // Create an empty window + viewController to host the Safari UI.
```

Comment:
Can "window + viewController" be replaced with "window & ViewController"? (Not sure if that's the intended meaning here)